### DISCUSSION TITLE
Hardware: suggestions, what's next?

### DISCUSSION THREAD

**Discussion by japuzzo** [shorepine/tulipcc#12] [General]
First off I love the project and want to promote it in the retro-computing circles. When I have discussed this project with others I keep getting these questions:

- Is there a VGA or HDMI connection so I can use old or un-used monitor?
- There is a USB connector for the keyboard but what about a joystick or game controller. Since this system seems perfect for creating new retro-games.
- Where can I get the hardware? At least a through hole kit with all the pieces ( add your own STM32-S3 )

My own personal question or suggestion is documenting what you want to do next with the project. This way other like myself who want to help know the direction and/or what's next.

**Comment by bwhitman** [comment]
I'd love to continue evolving Tulip once we have a core group of people using it / working on it. We're still incredibly early days so I think a bigger decision on Tulip's evolving specs may have to wait a bit. 

The original design goal of Tulip is something you can build on your own (or use the software simulator) with relatively cheap off the shelf (and in stock) parts in a **laptop form factor** -- that is, portable, with a portable and readable screen included. That limits some of the changes people might want. I'd also like to keep it to "one specification" - I don't want to support a more powerful Tulip alongside a less capable one.

Re: where can you get the hardware -- this should be relatively clear, I hope, [on the various build pages](https://github.com/bwhitman/tulipcc/blob/main/docs/tulip_build.md). If you're having trouble getting the parts, let us know here. 

So with that said
- VGA: certainly possible with [FabGL](https://github.com/fdivitto/fabgl) or otherwise. I don't think it would make a great laptop, and we'd be limiting our resolution to likely 640x480 at 6 bit color (I need to test this more) and the boards would have to add the resistor ladder, but I don't see why we couldn't adapt our GPU code to output to VGA as of right now. We'd lose multitouch, unless your VGA display had an capacitive touch i2c layer. But certainly as a separate output to the 40-pin RGB TTL, it wouldn't be a bad idea. 
- HDMI - I do not believe this is possible ever with the MCU we use. Certainly someone could build a TTL RGB -> HDMI capture converter using an FPGA but it won't be something we'd want to support in Tulip itself
- joystick: **yes**, I definitely want to add that. If USB, we would have to figure out how to support USB hubs in the firmware. Maybe an easier, and to be honest, more Tulip-y thing is to support SNES style latch/clock/data joysticks. Less parsing overhead and I think we have the pins free.

**Comment by jacquesgroenewald** [comment]
I love this project and cant wait to build one and test!

Currently how does one get external files onto the unit? Is it possible? Perhaps adding an SD card reader would be the answer?

Also out of curiosity, on the current unit, can one toggle the IOs in the Mircopython way (from machine import Pin, etc....)? I know there are not many IO's left but I wondered if it is possible to toggle them.

I dont suggest this becomes a "Raspberry Pi" type unit but to have even 3 or 4 IOs available would be nice

But then again how do you add an SD card if there are not enough IOs available...

**Comment by bwhitman** [comment]
You can get files on from a computer in a few ways:

 - Put the files on a web host or server and use `tulip.url_save(url, filename)` (you can do this easily with `python -m SimpleHttpServer` from the directory you want on your computer) 
 - Put your files in `tulip_home` in this repository and run `python tulip_fs_create.py` from `ports/esp32s3` to flash the filesystem, it will copy whatever is in there over (and overwrite anything already there)
 - Use [Micropython WebREPL](https://micropython.org/webrepl/)'s file transfer feature
 - (Soon!) Use **TULIP WORLD** to share the files from Tulip Desktop and access it on Tulip CC
 
An SD card is certainly possible (and there may be the pins for it!) But it's not on my personal roadmap -- I easily get by with one of the above options to get files over. 

Re: GPIOS, absolutely! All that stuff will work. There's currently 7 free GPIOs available on the ESP32S3, and also the I2C bus is available -- we just have the touchscreen on it now. You can easily get at them with the breadboard or breakout board versions of Tulip. The integrated board does not break them out (but could!) except for I2C.

**Reply by jacquesgroenewald** [reply]
Awesome!

I will try it out as soon as have built my own unit. Thanks

I also hope VGA would be possible as mentioned in the thread above. 

In the meantime, it may be of interest, but I found this commercially available RGB I/F to HDMI converter at quite a low cost: under $20
(but it has the 50way connector and not the 40way). I have not tested it.

https://www.myirtech.com/list.asp?id=664

**Reply by bwhitman** [reply]
That's a super cool part! I bet you could wire that up to Tulip pretty easily and get HDMI out, at least from a breadboard. Or, from an existing 40pin, could get two breakouts -- one 40 and one 50, and wire the signals using short jumper cables.
